quinn-proto/src/connection/assembler.rs

```diff
@@ -148,20 +131,72 @@ impl Assembler {
         if bytes.is_empty() {
             return;
         }
-
-        self.data.push(Buffer { offset, bytes });
-        // Why 32: on the one hand, we want to defragment rarely, ideally never
+        let buffer = Buffer::new(offset, bytes, allocation_size);
+        self.buffered += buffer.bytes.len();
+        self.allocated += buffer.allocation_size;
+        self.data.push(buffer);
+        // Rationale: on the one hand, we want to defragment rarely, ideally never
         // in non-pathological scenarios. However, a pathological or malicious
         // peer could send us one-byte frames, and since we use reference-counted
         // buffers in order to prevent copying, this could result in keeping a lot
-        // of memory allocated. In the worst case scenario of 32 1-byte chunks,
-        // each one from a ~1000-byte datagram, using 32 limits us to having a
-        // maximum pathological over-allocation of about 32k bytes.
-        if self.data.len() - self.defragmented > 32 {
+        // of memory allocated. This limits over-allocation in proportion to the
+        // buffered data. The constants are chosen somewhat arbitrarily and try to
+        // balance between defragmentation overhead and over-allocation.
+        let buffered = self.buffered.min((self.end - self.bytes_read) as usize);
+        let over_allocation = self.allocated - buffered;
```

Comment:
This is a invariant that must be maintained. I added comments to the `buffered` and `allocated` field stating this invariant.

Reply:
The only way the invariant might be invalidated is when we pass an `allocation_size` that is smaller than `bytes.len()` to `insert` in release builds, because we currently only use a `debug_assert!` to protect against that. Maybe we should change this back to an `assert!`?

Reply:
No, that should be fine.